File: tkpdseller/src/main/java/com/tokopedia/seller/product/edit/view/fragment/ProductAddFragment.java
package com.tokopedia.seller.product.edit.view.fragment;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.tokopedia.seller.product.common.di.component.ProductComponent;
import com.tokopedia.seller.product.edit.di.component.DaggerProductAddComponent;
import com.tokopedia.seller.product.edit.di.module.ProductAddModule;
import com.tokopedia.seller.product.edit.view.activity.ProductAddActivity;
import com.tokopedia.seller.product.edit.view.model.edit.ProductViewModel;
import com.tokopedia.seller.product.edit.view.model.upload.intdef.ProductStatus;
import com.tokopedia.seller.product.edit.view.presenter.ProductAddPresenter;

import java.util.ArrayList;

//import com.tokopedia.seller.product.edit.di.component.DaggerProductAddComponent;

/**
 * Created by nathan on 4/3/17.
 */

public class ProductAddFragment extends BaseProductAddEditFragment<ProductAddPresenter>  {

    /**
     * Url got from gallery or camera or other paths
     */
    protected ArrayList<String> imageUrlListFromArg;

    public static ProductAddFragment createInstance(ArrayList<String> tkpdImageUrls) {
        ProductAddFragment fragment = new ProductAddFragment();
        if (tkpdImageUrls != null && tkpdImageUrls.size() > 0) {
            Bundle bundle = new Bundle();
            bundle.putStringArrayList(ProductAddActivity.EXTRA_IMAGE_URLS, tkpdImageUrls);
            fragment.setArguments(bundle);
        }
        return fragment;
    }

    @Override
    public void initInjector() {
        DaggerProductAddComponent
                .builder()
                .productAddModule(new ProductAddModule())
                .productComponent(getComponent(ProductComponent.class))
                .build()
                .inject(this);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        Bundle args = getArguments();
        if (args != null && args.containsKey(ProductAddActivity.EXTRA_IMAGE_URLS)) {
            imageUrlListFromArg = args.getStringArrayList(ProductAddActivity.EXTRA_IMAGE_URLS);
        }
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        saveDefaultModel();
        return view;
    }

    @Override
    public void onSuccessLoadShopInfo(boolean isGoldMerchant, boolean isFreeReturn, boolean officialStore) {
        super.onSuccessLoadShopInfo(isGoldMerchant, isFreeReturn, officialStore);
        saveDefaultModel();
    }

    private void saveDefaultModel(){
        //save default value here, so we can compare when we want to save draft
        // will be overriden when not adding product
        firstTimeViewModel = collectDataFromView();
    }

    @Override
    public boolean showDialogSaveDraftOnBack(){
        // check if this fragment has any data
        // will compare will the default value and the current value
        // if there is the difference, then assume that the data has been added.
        // will be overriden when not adding product
        ProductViewModel model = collectDataFromView();
        return !model.equals(firstTimeViewModel);
    }

    @Override
    public boolean isNeedGetCategoryRecommendation() {
        return true;
    }

    @ProductStatus
    public int getStatusUpload() {
        return ProductStatus.ADD;
    }

}

File: features/merchant/product_manage_item/src/main/java/com/tokopedia/product/edit/view/fragment/ProductAddFragment.kt
package com.tokopedia.product.edit.view.fragment

import android.os.Bundle
import com.tokopedia.product.edit.common.di.component.ProductComponent
import com.tokopedia.product.edit.common.util.ProductStatus
import com.tokopedia.product.edit.di.component.DaggerProductAddComponent
import com.tokopedia.product.edit.di.module.ProductAddModule
import com.tokopedia.product.edit.price.model.ProductCatalog
import com.tokopedia.product.edit.price.model.ProductCategory
import com.tokopedia.product.edit.price.model.ProductName
import com.tokopedia.product.edit.utils.convertImageListResult
import com.tokopedia.product.edit.utils.isFilledAny
import com.tokopedia.product.edit.view.listener.ProductAddView
import com.tokopedia.product.edit.view.model.ProductAddViewModel
import com.tokopedia.product.edit.view.presenter.ProductAddPresenterImpl
import java.util.ArrayList

class ProductAddFragment : BaseProductAddEditFragment<ProductAddPresenterImpl<ProductAddView>, ProductAddView>() {
    override var statusUpload: Int = ProductStatus.ADD

    override fun initInjector() {
        DaggerProductAddComponent
                .builder()
                .productAddModule(ProductAddModule())
                .productComponent(getComponent(ProductComponent::class.java))
                .build()
                .inject(this)
        presenter.attachView(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        presenter.attachView(this)
        if(currentProductAddViewModel == null){
            currentProductAddViewModel = ProductAddViewModel()
        }
        currentProductAddViewModel?.productCatalog = arguments?.get(EXTRA_CATALOG) as ProductCatalog?
        currentProductAddViewModel?.productCategory = arguments?.get(EXTRA_CATEGORY) as ProductCategory?
        currentProductAddViewModel?.productName = arguments?.get(EXTRA_NAME) as ProductName?
        currentProductAddViewModel?.productPictureList = arguments?.getStringArrayList(EXTRA_IMAGES)?.convertImageListResult()
    }

    companion object {
        fun createInstance(productCatalog: ProductCatalog?, productCategory: ProductCategory?, productName: ProductName?, productImages: ArrayList<String>?) : ProductAddFragment{
            val productAddFragment = ProductAddFragment()
            val bundle = Bundle()
            bundle.putParcelable(EXTRA_CATALOG, productCatalog)
            bundle.putParcelable(EXTRA_NAME, productName)
            bundle.putParcelable(EXTRA_CATEGORY, productCategory)
            bundle.putStringArrayList(EXTRA_IMAGES, productImages)
            productAddFragment.arguments =bundle
            return productAddFragment
        }
    }

    override fun showDialogSaveDraftOnBack(): Boolean {
        // check if this fragment has any data
        // will compare will the default value and the current value
        // if there is the difference, then assume that the data has been added.
        // will be overriden when not adding product
        return currentProductAddViewModel?.isFilledAny()?:false
    }
}